File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
int target;
bool solve(int index,int currsum,vector<int>&nums,vector<vector<int>>&dp){
    if(currsum==target){
        return true;
    }
    if(index==nums.size()){
        return false;
    }
    if(currsum>target){
        return false;
    }
    if(dp[index][currsum]!=-1){
        return dp[index][currsum];
    }
    bool take=solve(index+1,currsum+nums[index],nums,dp);
    bool nottake=solve(index+1,currsum,nums,dp);
  return dp[index][currsum]=(take||nottake);
}
    bool canPartition(vector<int>& nums) {
       int totalsum=0;
        int n=nums.size();
       for(int it:nums){
        totalsum+=it;
       }
       if(totalsum%2!=0){
         return false;
       }
       target=totalsum/2;
       vector<vector<int>>dp(n+1,vector<int>(target+1,-1));

       return solve(0,0,nums,dp);

    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    boolean solve(int i, int n, int[] nums, int target, int[][] dp) {

        if (target == 0) {
            return true;
        }

        if (i == n || target < 0) {
            return false;
        }

        if (dp[i][target] != -1) {
            return dp[i][target] == 1;
        }

        boolean take = solve(i + 1, n, nums, target - nums[i], dp);

        boolean nottake = solve(i + 1, n, nums, target, dp);

        dp[i][target] = (take || nottake) ? 1 : 0;

        return take || nottake;
    }

    public boolean canPartition(int[] nums) {
        int sum=0;
        for(int i=0;i<nums.length;i++){
            sum+=nums[i];
        }
        if(sum%2!=0){
            return false;
        }
        int target=sum/2;
        int [][]dp=new int[nums.length+1][target+1];
        for(int i=0;i<=nums.length;i++){
            Arrays.fill(dp[i],-1);
        }
        return solve(0,nums.length,nums,target,dp);
    }
}